Camera Culling Mask (Layers)

tsone opened the pull request:
Answers feature request #5562 , but with bare-bones functionality (no bitmask/field API). The implementation here should work with lights, cubemap camera and shadow maps. Please inform me if there's some functionality I missed that would benefit from this.

The new example `webgl_shadowmap_layerbits` demonstrates the usage with shadow maps. I also modified the example `webgl_materials_cubemap_dynamic2` to demostrate the use with `CubeCamera`. (The sphere is now hidden using layers.) No other examples were modified.

The new `Camera.cullingMask` property is a bitmask where the set bits indicate visible layers to the camera. Unset bits mark invisible (=hidden, culled) layers. Default value is `0xFFFFFFFF`, i.e. all layers are visible. The new `Object3D.layerBits` property is a bitfield where the set bits indicates the layers where object is on. Object can be on multiple layers. Default value is `0xFFFFFFFF`, i.e. object is on all layers.

An object (and its branch) is culled if `( object.layerBits & camera.cullingMask ) === 0`. This culling affects **all kinds of objects in the scene graph** -- be it `Scene`, `Light`, `Mesh`, `Camera`, etc. It does not affect the scene hierarchical transformations, but it does for example affect light allocation.

The actual culling test is done in the function `Object3D.isVisible( camera )`. In practice, what I did was to replace all `object.visible` checks with this function. In my opinion all of the checks I found were regards to some camera, so I think the function could have as well been in `Camera` instead of the `Object3D` (i.e. `Camera.isVisible( object )`).

In `WebGLRenderer`, I needed to use the `_currentCamera` quite extensively to cull the lights properly. I needed to set this more early than it was (in `setProgram()`) because so much depended on it. I am not 100% sure if this is correct, but it seems so.

Please comment. One thing I'm not sure is if the name `layerBits` is actually good as Unity uses `layerMask` and this feature is modeled after Unity. Which one is better?

WestLangley commented:
Very well-executed and well-written demo, IMHO. : - )

I agree, there is no need to implement a different API until we agree on what we are doing.

I am having a hard time getting my head around the concept of having lights in layers, though.

---

As an aside, is there a way to see the demo without having to jump through these hoops?

    git checkout -b dev-test
    git pull git://github.com/user_name/three.js.git branch_name

tsone commented:
The new demo is mainly cut-paste from the old one, so don't thank me :)

This PR ___does not___ enable lights to illuminate only objects in the same layers, i.e. _selective lighting_ #5180. One could however achieve requested result by reserving a layer for each room and its primary lights, and then calling `render()` for each room/layer with its culling mask bit on. It's not the same of course, as the requester wanted a single `render()` to do the magic.

This PR is about camera culling. Lights will be also culled (as everything else) if they are not on any of the masked layers. About shadow mapping, as you know, it works by having a _shadow camera_ and rendering through it, so this PR works directly with that. I made it so that light's layer bits are used directly as culling mask for the shadow camera, which means only objects in the same layers as the light will be shadow casters. The new example demonstrates this. Or do you propose other way?

---

I don't have the demo online anywhere right now, so to test you need to pull. (Have a script that will make testing PRs easier?)

WestLangley commented:
I prefer the name `layerMask` to `layerBits`. We will add a friendly API later, anyway.

Here is a simpler proposal.

1. Remove `cullingMask`.
2. Every object has a `layerMask`.

For a camera, `layerMask` identifies the layers the camera can _see_. For a light, `layerMask` identifies the layers the light _illuminates_. For renderable objects, it identifies the layers the object is _in_.

For non-renderable objects like `Object3D`, it does not apply.

The property is not inherited.

Optional. If the camera is a _shadow camera_, `layerMask` identifies the layers in which the shadow camera's light casts shadows. Personally, I think this would a bit complicated, so I would not recommend it. There is not a one-to-one correspondence between shadow cameras and lights when lights are cascading, anyway.

---

Script helps. Good idea.

tsone commented:
I will change to `layerMask` as this is more familiar to Unity users.

I don't think it's a good idea to have `layerMask` to serve many purposes depending of the object type, like replacing `cullingMask` for `Camera`. Such is confusing and probably complicates the code. The name `cullingMask` is from #5562 and there are no complaints against it. The name states its purpose quite well and it also matches with Unity. Can you clarify more why change `cullingMask`?

> For a light, layerMask identifies the layers the light illuminates.

I think so also, but I didn't implement this as it's more complicated. This can be added later.

> If the camera is a shadow camera, layerMask identifies the layers in which the shadow camera's light casts shadows. 

This is in the PR. Like I explained above, light's `layerMask` property is used directly as the shadow camera's `cullingMask`. This goes with the logic that the light illuminates the layers it is on.

WestLangley commented:
> I don't think it's a good idea to have layerMask to serve many purposes depending on the object type

That is true.  However, in my proposal (which was a trial-ballon, BTW), the `layerMask` just identifies what layer the object is "in". I was describing what being "in a layer" means for various objects. The proposal is simpler than what is proposed in this PR.

> Like I explained above, light's layerMask property is used directly as the shadow camera's cullingMask. 

And what is the shadow camera's `layerMask` used for in this PR?

Why does it make sense for a camera to have both a `layerMask` and a `cullingMask`?

tsone commented:
Originally this feature was about culling objects in the scene graph in respect to a camera as per #5562 and to help in my Soft Particles example #5911, but now it seems selective lighting #5180 is creeping in... I would like to keep culling and lighting separate.

> The proposal is simpler than what is proposed in this PR.

I'm not sure what proposal do you mean -- or if I really understand at all :) Can you tell a bit more how it would be simpler (to implement)?

> Why does it make sense for a camera to have both a layerMask and a cullingMask?

The reason is that `Camera` can have children and this PR culls entire branches instead of single objects. The culling behavior is similar to setting 'Object3D.visible'. Culling and lighting are also different things and would probably benefit from different masks.

I understand culling by material would be good for selective lighting as it would be probably easier to manage the light uniform updating -- and would be probably more efficient. I'm not however implementing selective lighting here. That would be an another feature.

WestLangley commented:
> I'm not sure what proposal do you mean -- or if I really understand at all :)

1. Remove `cullingMask`.
2. Every object has a `layerMask` -- cameras and lights included. The `layerMask` is not inherited.

> Can you tell a bit more how it would be simpler (to implement)?

It is not simpler to implement. It is simpler to understand.

> The reason is that Camera can have children and this PR culls entire branches instead of single objects. The culling behavior is similar to setting Object3D.visible. 

Right. And that is too confusing, IMHO.

> An object (and its branch) is culled if ( object.layerBits & camera.cullingMask ) === 0. 

In my proposal, entire branches are never culled based on the layer -- an object's layer has no effect on its children. That is what `object.visible` is for.

That being said, my proposal is just a straw-man. Feel free to shoot it down if you want. :-)

tsone commented:
Yup. I got your point and I think it's a good idea. Although I would like to keep `cullingMask` -- but I can go with your proposal.

Did I get you right your proposal would deal with the following sections in `WebGLRenderer` that deal with `material.visible`:

```javascript
this.renderBuffer = function ( camera, lights, fog, material, geometryGroup, object ) {

    if ( material.visible === false ) return;

    updateObject( object );

    var program = setProgram( camera, lights, fog, material, object );
    ...
```

Right? I cull here if `object.layerMask` doesn't match camera's. Do I also "strip" lights from the lights array? I didn't look into this yet. What's your impression?

WestLangley commented:
OK, then. Let's think about another approach that allows you to keep `cullingMask`.

It doesn't make sense for non-renderable objects to have a mask if masks are not inherited. So here is Proposal 2:

1. Every _renderable_ object has a `layerMask`.
2. Every light has a `layerMask`.
3. Every camera has a `cullingMask` (no `layerMask`).

Consequently, `Object3D`, itself, has no mask property. Also, the masks are not inherited.

The one thing I am not sure I like about the term `cullingMask`, however, is if a bit is 1, does that mean the layer is _culled_? I don't think that is what you intended. I think you intended the opposite.

I have not thought about the code changes required to implement this. I am more interested right now in agreeing to the functionality.

tsone commented:
I agree with your new proposal and it would be nice, but the problem is that there is no `Renderable` class. Should it be introduced? This could be done by pulling out some of the "renderables-related" functionality from `Object3D`. This is of course risky as there are no unit tests... Anyway, this proposal makes a lot more sense otherwise.

My idea was that bit 1 in `cullingMask` means the layer is visible, 0 means it's invisible (=culled). I'm not sure how Unity has the bits set. This allows the visiblity check to be a single bitwise AND:

```javascript
if ( ( renderable.layerMask & camera.cullingMask ) === 0 ) {
    not_visible();
} else {
    visible();
}
```

This is very cheap. If the bits in `cullingMask` mean the opposite, then bits must be inverted. I'd set both `layerMask` and `cullingMask` initially to `0xffffffff` (=all layers enabled), as that is the default behavior before this layers functionality.

BTW. This was probably mentioned by someone before, but another way to have layers is to use _layer names_ as an array in the renderable:
```javascript
levelWall.layers = [ 'level', 'opaque' ];
levelWindowPane.layers = [ 'level', 'opaque' ];
levelWindowGlass.layers = [ 'level', 'transparent' ];

playerHead.layers = [ 'player', 'opaque' ];
```
Problem here is the visibility check that must check if strings exists -- which it's must more expensive. I just mention this so you see I've thought about this option too.

tsone commented:
Nevermind the previous about `Renderable` class -- it's not necessary to have that.

tsone commented on src/renderers/webgl/plugins/ShadowMapPlugin.js:
```diff
@@ -318,7 +320,7 @@ THREE.ShadowMapPlugin = function ( _renderer, _lights, _webglObjects, _webglObje
 				object = webglObject.object;
 
-				if ( object.visible && object.castShadow ) {
+				if ( object.isVisible( shadowCamera ) && object.castShadow ) {
```
Should I have the camera layers check here? This loop is about rendering immediate objects. This already checks if object casts shadows, so my guess is the layers check is needed, right?

tsone commented:
Modified PR so that it matches more on @WestLangley's proposal. `Object3D.layerMask` is used as camera's culling mask, so `cullingMask` is removed. I take back what I said earlier.

__Do not merge this yet!__ I decided to give a try on the selective lighting as well (=lights illuminate objects in the same layers). My current idea would be to selectively enable/disable lights at the uniform level when the object is rendered. Any help locating the meaningful bits to do this in the `WebGLRenderer`? Is this a good way?  Any suggestions? :smile_cat:

WestLangley commented:
> Any help locating the meaningful bits to do this in the WebGLRenderer?

I think `WebGLRenderer` may need some refactoring, but I expect the easiest thing to do for now is to pass 0 for the light's intensity into `setColorLinear()`, which is called from `setupLights()`, r.71dev.

Commit:
Added support for layers (layerMask) and selective lighting (and shadows). #5562 #5180

tsone commented:
Now it should be done. I collapsed the commits into one and rewrote the commit history. This adds layer mask support where only the objects on the same layers as camera are rendered (resolves #5562), and lights (and their shadow maps) only affect the objects on the same layers (resolves #5180).

I also added a second light in the `webgl_shadowmap_layermask` example to show this works with multiple lights. Play around with the masks, for example see what happens when all light bits are 0. (Note, background will remain visible because it has default layer mask and camera mask has bits 31-4 set.)

As earlier, `webgl_materials_cubemap_dynamic2` was modified to use layers. This shows layers also work with the cube camera.

tsone commented:
Forgot to say, in `WebGLRenderer.setProgram()`, light uniforms are updated by running `setupLights()` every time when the resulf of `camera.layerMask & object.layerMask` changes in `setProgram()` -- this in addition to previous conditions to update light and/or material uniforms.

Shadow uniforms are also updated on the same conditions. However, shadows can't be disabled after shader compilation, so as workaround `shadowDarkness` is set to 0 to hide the shadow effect.

WestLangley commented:
This is a fine piece of work. : - )

tsone commented:
Thanks :) It's a big change affecting critical parts. If you see any problem, please let me know.

tsone commented:
@Zob1 I don't know why it's not merged. There's no explanation. Certainly you could merge this to your own fork.

mrdoob commented:
I wanted to solve other parts of the library before having a look at this. However, this implementation is too intimidating. Users shouldn't have to know what bitwise operations are.

WestLangley commented:
As I said earlier in this post, we can add a user-friendly API on top of this.

The bit-mask concept is simple:

> For a camera, `layerMask` identifies the layers the camera can see. For a light, `layerMask` identifies the layers the light illuminates. For renderable objects, it identifies the layers the object is in.

> For non-renderable objects like Object3D, it does not apply.

> The property is not inherited.

For a previous discussion regarding a user-friendly API see https://github.com/mrdoob/three.js/issues/5562.

mrdoob commented:
Also, I would start with camera/objects first. Lights is a trickier one.

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#5931

tsone commented:
What was the problem with this pr?

mrdoob commented:
The API wasn't very user friendly. The current API is `.layers.set(1)` or `.layers.enable(1)`. The user doesn't need to know about bitwise operations.

I used your PR as reference though. So thanks! 😊

tsone commented:
No problem! Thank you!
